### PR TITLE
Prevent silent (ish) failure of agent validation

### LINKED ISSUE #1824: Link from AGENT_ADDRESS not working
## Issue
Linking to an agents `AGENT_ADDRESS` doesn't seem to propagate to another conductor.
## Steps to reproduce
Alice and Bob are separate agents running on separate `holochain` conductors pointing to the same local sim2h. (Although the same behavior happens with n3h). 
1. Alice commits an entry and links it `from` her `AGENT_ADDRESS`.
2. Alice is able to retrieve this entry by getting the link.
3. Bob tries to get the same entry using Alice's `AGENT_ADDRESS`.
4. Bob gets back an empty list even after waiting for 15 minutes.

## Source
- DNA https://gist.github.com/freesig/23e15c7e41167fc3d353e4eaafd98de1
- Alice's config https://gist.github.com/freesig/89046a1641db4e983a6f65b70cb03301
- Bob's config https://gist.github.com/freesig/4370676c0d22c11694656f7dc0cbf099

## Debug
I think this might be the issue. __could not be validated due to unresolved dependencies.__
This happens when Bob tries to get the links in step 3.
```DEBUG 2019-11-01 15:22:20 [holochain::test-instance] retry_validation/puid-d-ed crates/core/src/workflows/hold_link.rs:78 workflow/hold_link: Link could not be validated due to unresolved dependencies and will be tried later. List of missing dependencies: [HashString("HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez"), HashString("QmQHBpnU5f9zccWx7vNRnWVMvzd91aFinXPdis7dCspcT6")]
DEBUG 2019-11-01 15:22:20 [holochain::test-instance] retry_validation/puid-d-ee crates/core/src/workflows/hold_entry.rs:71 workflow/hold_entry: QmfTXMcaHgfTZ9pfMJWCsDrtpHqu4d2v7ED29y5wWxaJ7C could not be validated due to unresolved dependencies and will be tried later. List of missing dependencies: [HashString("HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez"), HashString("QmQHBpnU5f9zccWx7vNRnWVMvzd91aFinXPdis7dCspcT6")]
```
This might be related. It happens at the beginning.
```
DEBUG 2019-11-01 15:19:17 [holochain::test-instance] store_entry_content/puid-4-4a crates/core/src/workflows/hold_entry.rs:91 workflow/hold_entry: is valid! Qma8DJeNaoueNUdTw3TsVkjDzEnr7zDegxgZ3WRwrdyQPY
DEBUG 2019-11-01 15:19:17 [holochain::test-instance] store_entry_content/puid-4-4b crates/core/src/workflows/hold_entry.rs:91 workflow/hold_entry: is valid! QmTXDBpRRQfHKuiGPi41j5zCsfvbv9GfvrD7MhWbcLHAMY
WARN 2019-11-01 15:19:17 [holochain::test-instance] store_entry_content/puid-4-49 crates/core/src/workflows/hold_entry.rs:83 workflow/hold_entry: Entry HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez is NOT valid! Validation error: Error(ErrorGeneric("Failed to validate agent ID on a zome, [Error(ErrorGeneric(\"I/O Error: UnexpectedEof\"))]"))
ERROR 2019-11-01 15:19:17 [holochain::test-instance] store_entry_content/puid-4-49 crates/core/src/network/handler/store.rs:38 net/dht: Failed to validate agent ID on a zome, [Error(ErrorGeneric("I/O Error: UnexpectedEof"))]
```

## Alice
This is Alices DHT at the time when Bob tries to get the links.
```
Dht:
====
Holding:
* [UNKNOWN] QmNzgmwB6NR7wGsLH3FjXnj5sDj8CWos9rN9MxkTXqX4ev: {"entry_type":"CapTokenGrant","entry_address":"QmWXw8U8j8r6Ex8KJ4NDAmwr5iZdAZKgR3S7o4D8GC65GD","provenances":[["HcSCj4uMm999rT4B6kfgSYx6ONfg3misvoV76JI9J57KM89ejVf4uwhm7Mm6f7i","Zsld37WwMiDOk8xa5ahZJ1pRL8f1zkAYax91AMUhvgI8liglDDYyYddMlNt3jOf6J2c+cZ8g4uAbNj1VlllrCg=="]],"link":"Qmbay3GB4nemC8SDSvVZhQFmnCCsxWxBWTkhYCxVawojmw","link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:19:16+00:00"}
* [UNKNOWN] QmWFezMP8d7EVtHkLDWi4Xh5oTkapTQPiFicf9Z2r3UuNr: {"entry_type":"Dna","entry_address":"Qme9YzqCvBgywyE4X8in5QZkP6L4CNfWq8zhwoJ2ktjB5U","provenances":[["HcSCj4uMm999rT4B6kfgSYx6ONfg3misvoV76JI9J57KM89ejVf4uwhm7Mm6f7i","uXg7iTpDQI7rR8YmaKbYZKwqgSFjxqy5tZ8eI9CGASMnDRaFfDRJi7o7IgHCs3/0nsO9aCqYIMxXp3J9CFtSCA=="]],"link":null,"link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:19:16+00:00"}
* [UNKNOWN] Qmbay3GB4nemC8SDSvVZhQFmnCCsxWxBWTkhYCxVawojmw: {"entry_type":"AgentId","entry_address":"HcSCj4uMm999rT4B6kfgSYx6ONfg3misvoV76JI9J57KM89ejVf4uwhm7Mm6f7i","provenances":[["HcSCj4uMm999rT4B6kfgSYx6ONfg3misvoV76JI9J57KM89ejVf4uwhm7Mm6f7i","fUwCL7JmiP3QC0Clj+qHoc9jNo0YMt2QxXLvPdpwXq7DpdmRXMi86gPmZqOZ+x///TNy/Ifxg6BZ/CaJLsQZBA=="]],"link":"QmWFezMP8d7EVtHkLDWi4Xh5oTkapTQPiFicf9Z2r3UuNr","link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:19:16+00:00"}
* [UNKNOWN] QmcU7BbBysfrMbJjUmrqa6QwqkFtNwuB8FMyhtuFWfYLYk: {"entry_type":{"App":"post"},"entry_address":"QmQHBpnU5f9zccWx7vNRnWVMvzd91aFinXPdis7dCspcT6","provenances":[["HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez","iwYa/8PpEAerWQ0drLVPbDYH0Bs5cJ9N4ouNEk4riCnzcDnBI6GWJ9t9FRuJ0iosdzvGP2oxv4GnBf41Kx5HAw=="]],"link":"QmTXDBpRRQfHKuiGPi41j5zCsfvbv9GfvrD7MhWbcLHAMY","link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:21:15+00:00"}
* [post] QmQHBpnU5f9zccWx7vNRnWVMvzd91aFinXPdis7dCspcT6: {"message":"I have a lot to say :)","timestamp":1572582075337,"author_id":"HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez"}
* [UNKNOWN] QmfUWrp6vXjrZTtBJNjEZJcBzRrcBVu6v9JingwYScTqwX: {"entry_type":"LinkAdd","entry_address":"QmfTXMcaHgfTZ9pfMJWCsDrtpHqu4d2v7ED29y5wWxaJ7C","provenances":[["HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez","ldkyuleCKCIQC+TJ2u9BAT5no0T2+Na/drKgcl7q7bjcL2HxqqICLH2mfT4EE7isj6R3EimyPaWlQPfhGOb0Aw=="]],"link":"QmcU7BbBysfrMbJjUmrqa6QwqkFtNwuB8FMyhtuFWfYLYk","link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:21:15+00:00"}
* [%link_add] QmfTXMcaHgfTZ9pfMJWCsDrtpHqu4d2v7ED29y5wWxaJ7C: (author_post#)
	HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez => QmQHBpnU5f9zccWx7vNRnWVMvzd91aFinXPdis7dCspcT6
* [%link_add] QmfTXMcaHgfTZ9pfMJWCsDrtpHqu4d2v7ED29y5wWxaJ7C: (author_post#)
	HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez => QmQHBpnU5f9zccWx7vNRnWVMvzd91aFinXPdis7dCspcT6
--------
```

## Bob
This is Bob's DHT at the time that he tries to get the links.
```
Dht:
====
Holding:
* [UNKNOWN] QmXJVGPm9WNJuyNUu5ysXFdhpekifcTw7Zr42WioM6fmFP: {"entry_type":"Dna","entry_address":"Qme9YzqCvBgywyE4X8in5QZkP6L4CNfWq8zhwoJ2ktjB5U","provenances":[["HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez","mbSas0wuxkIbUU2oFI+EbpqAPMDUvu+g899famamYKnB/2nWycW3BDx5MwhNNmlvGb0WREJni2IXatQ4t7CCDA=="]],"link":null,"link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:17:52+00:00"}
* [UNKNOWN] Qma8DJeNaoueNUdTw3TsVkjDzEnr7zDegxgZ3WRwrdyQPY: {"entry_type":"AgentId","entry_address":"HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez","provenances":[["HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez","7WMx4vshU8PVZhQblRTw0nRnVE89mycEJHxEIOWq6ENXmY87KCLjC3jppTybKMci57GjZhNAP3bZwhD977vCDQ=="]],"link":"QmXJVGPm9WNJuyNUu5ysXFdhpekifcTw7Zr42WioM6fmFP","link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:17:52+00:00"}
* [UNKNOWN] QmTXDBpRRQfHKuiGPi41j5zCsfvbv9GfvrD7MhWbcLHAMY: {"entry_type":"CapTokenGrant","entry_address":"QmWXw8U8j8r6Ex8KJ4NDAmwr5iZdAZKgR3S7o4D8GC65GD","provenances":[["HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez","62p6nLQEvlWGw4VqGlqVgvzj9R/TvCFxHMv7r5cvBTom3Zg111cTY4dk6HGnn1QuRgIbkqxwEpjRzwPjKUCxAw=="]],"link":"Qma8DJeNaoueNUdTw3TsVkjDzEnr7zDegxgZ3WRwrdyQPY","link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:17:52+00:00"}
* [UNKNOWN] QmcU7BbBysfrMbJjUmrqa6QwqkFtNwuB8FMyhtuFWfYLYk: {"entry_type":{"App":"post"},"entry_address":"QmQHBpnU5f9zccWx7vNRnWVMvzd91aFinXPdis7dCspcT6","provenances":[["HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez","iwYa/8PpEAerWQ0drLVPbDYH0Bs5cJ9N4ouNEk4riCnzcDnBI6GWJ9t9FRuJ0iosdzvGP2oxv4GnBf41Kx5HAw=="]],"link":"QmTXDBpRRQfHKuiGPi41j5zCsfvbv9GfvrD7MhWbcLHAMY","link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:21:15+00:00"}
* [post] QmQHBpnU5f9zccWx7vNRnWVMvzd91aFinXPdis7dCspcT6: {"message":"I have a lot to say :)","timestamp":1572582075337,"author_id":"HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez"}
* [UNKNOWN] QmfUWrp6vXjrZTtBJNjEZJcBzRrcBVu6v9JingwYScTqwX: {"entry_type":"LinkAdd","entry_address":"QmfTXMcaHgfTZ9pfMJWCsDrtpHqu4d2v7ED29y5wWxaJ7C","provenances":[["HcScjdwyq86W3w5y3935jKTcs4x9H9Pev898Ui5J36Sr7TUzoRjMhoNb9fikqez","ldkyuleCKCIQC+TJ2u9BAT5no0T2+Na/drKgcl7q7bjcL2HxqqICLH2mfT4EE7isj6R3EimyPaWlQPfhGOb0Aw=="]],"link":"QmcU7BbBysfrMbJjUmrqa6QwqkFtNwuB8FMyhtuFWfYLYk","link_same_type":null,"link_update_delete":null,"timestamp":"2019-11-01T04:21:15+00:00"}
--------
```
Bob also has these pending validations.
```
Pending validations:
<HoldLinkWorkflow> [UNKNOWN] QmfTXMcaHgfTZ9pfMJWCsDrtpHqu4d2v7ED29y5wWxaJ7C: Error trying to get type/content: The requested entry could not be found locally
<HoldEntryWorkflow> [UNKNOWN] QmfTXMcaHgfTZ9pfMJWCsDrtpHqu4d2v7ED29y5wWxaJ7C: Error trying to get type/content: The requested entry could not be found locally
```

### PR DESCRIPTION
**Pull request by willemolding**
Closes #1824 

# Issue description

This issue stemmed from artifacts in the .dna.json file. As the `validate_agent` callback must be run for all zomes in a DNA it iterates over everything in the `zomes` field. If there is a __META.. field in the zome object it will include this and try to use it to run the validation callback.

Instead of producing a proper error it just assumed the failure of validation of an agent and this is why links could not be retrieved off their address.

# Fix

This adds several fixes to prevent this bug
- [x] Validation callbacks will explicitly panic if an unknown error occurs. This seems more appropriate than assuming all errors mean a validation failure (in this case it is a corrupt DNA failure)

- [x] Adds a check on loading a .dna.json file that all the `code` fields in zomes contain some code (in future check they all have the required callbacks)

- [x] Update the `hc package` tool to not dump __META stuff in json by default. `--strip-meta` flag is now `--include-meta.` 

## testing/benchmarking notes

( if any manual testing or benchmarking was/should be done, add notes and/or screenshots here )

## followups

( any new tickets/concerns that were discovered or created during this work but aren't in scope for review here )

## changelog

- [ ] if this is a code change that effects some consumer (e.g. zome developers) of holochain core,  then it has been added to [our between-release changelog](https://github.com/holochain/holochain-rust/blob/develop/CHANGELOG-UNRELEASED.md) with the format 

```markdown
- summary of change [PR#1234](https://github.com/holochain/holochain-rust/pull/1234)
```

## documentation

- [ ] this code has been documented according to our [docs checklist](https://hackmd.io/@freesig/Hk9AmKJNS)

### CODE REVIEW

File: crates/core_types/src/dna/mod.rs [hunk 1]
```diff
@@ -257,6 +257,38 @@ impl Dna {
...
+                } else {
+                    Err(HolochainError::ErrorGeneric(format!(
+                        "Zome {} has no code!",
+                        zome_name
+                    )))
+                }
+            })
+            .filter_map(|r| match r {
+                Ok(_) => None,
+                Err(e) => Some(e.to_owned()),
```
Comment (AshantiMutinta) [inline crates/core_types/src/dna/mod.rs]: r.err() will return an option so we could probably change this to filter_map(|r| r.err())

File: crates/core_types/src/dna/mod.rs [hunk 2]
```diff
@@ -257,6 +257,38 @@ impl Dna {
...
+            })
+            .collect();
+        if errors.len() == 0 {
```
Comment (AshantiMutinta) [inline crates/core_types/src/dna/mod.rs]: errors.is_empty()

File: crates/core_types/src/dna/mod.rs (hunk 1 above)
Comment (willemolding) [reply crates/core_types/src/dna/mod.rs]: Nice!

File: crates/core_types/src/dna/mod.rs (hunk 2 above)
Comment (willemolding) [reply crates/core_types/src/dna/mod.rs]: :+1:

**Pull request merged**